### PR TITLE
Fix LruCache. Make TreeCache track its own size.

### PR DESCRIPTION
**Pull request by erikjohnston**
The `LruCache` failed to reset `self.size` when `clear`  was called. This meant that `self.size` could be much large than the actual size, thereby reducing the actual effective size of the cache.

This PR fixes the issue by reverting the `LruCache` to not track size and instead use `len(..)` on the underlying cache. `TreeCache` now keeps track of its own size.

### CODE REVIEW

**Comment by NegativeMjark** [comment]
LGTM

File: synapse/util/caches/lrucache.py [hunk 1]
```diff
@@ -142,10 +140,11 @@ def cache_clear():
             list_root[NEXT] = list_root
             list_root[PREV] = list_root
             cache.clear()
+            self.size = 0
```
Comment (NegativeMjark) [inline synapse/util/caches/lrucache.py]: Why is this still here?

File: synapse/util/caches/treecache.py [hunk 2]
```diff
@@ -57,4 +60,33 @@ def pop(self, key, default=None):
                 break
             node_and_keys[i+1][0].pop(k)
 
+        popped, cnt = _strip_and_count_entires(popped)
+        self.size -= cnt
         return popped
```
Comment (NegativeMjark) [inline synapse/util/caches/treecache.py]: If you are going to enumerate the popped tree here then you might as well return a list of popped entries for the LruCache to remove from its linked list.

File: synapse/util/caches/treecache.py [hunk 3]
```diff
@@ -57,4 +60,33 @@ def pop(self, key, default=None):
...
+
+    def __len__(self):
+        return self.size
+
+
+class _Entry(object):
+    __slots__ = ["value"]
+
+    def __init__(self, value):
+        object.__setattr__(self, "value", value)
```
Comment (NegativeMjark) [inline synapse/util/caches/treecache.py]: Why not just ``self.value = value``?

File: synapse/util/caches/lrucache.py (hunk 1 above)
Comment (NegativeMjark) [inline synapse/util/caches/lrucache.py]: Also why didn't this fix the problem OOI?

**Comment by NegativeMjark** [comment]
Also could you add a test to check you've fixed the bug?

File: synapse/util/caches/lrucache.py (hunk 1 above)
Comment (erikjohnston) [inline synapse/util/caches/lrucache.py]: I dislike the `LruCache` trying to keep track of the size of the underlying cache. I much prefer that caches/collections keep track of their own size.

File: synapse/util/caches/treecache.py (hunk 2 above)
Comment (erikjohnston) [inline synapse/util/caches/treecache.py]: I'd quite like to keep the `pop` method returning a single item when given a full key, as that matches expectations.

**Comment by erikjohnston** [comment]
@NegativeMjark PTAL

**Comment by NegativeMjark** [comment]
LGTM

**Pull request merged**